File: offer/JZ36 二叉搜索树与双向链表/tree.java
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    private TreeNode head = new TreeNode(0);
    private TreeNode tail = head;
    private void pre(TreeNode root){
        if(root != null){
            pre(root.left);
            TreeNode prev = tail;
            tail.right = new TreeNode(root.val);
            tail = tail.right;
            tail.left = prev;
            pre(root.right);
        }
    }
    public TreeNode Convert(TreeNode root) {
        if(root == null)
            return null;
        pre(root);
        head = head.right;
        head.left = null;
        return head;
    }
}

File: 0094.BinaryTreInorderTraversal/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    vector<int> inorderTraversal(TreeNode* root) {
        if(root == nullptr)
            return {};
        vector<int> ret;
        stack<TreeNode*> stk;
        TreeNode *head = root;
        while(head != nullptr || !stk.empty()){
            while(head != nullptr){
                stk.push(head);
                head = head->left;
            }
            if(!stk.empty()){
                TreeNode* top = stk.top();
                stk.pop();
                ret.push_back(top->val);
                head = top->right;
            }
        }
        return ret;
    }
};